### PR TITLE
[DNM] refine `handleJoin`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4118

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 1]
```diff
@@ -464,101 +462,32 @@ ExpressionActionsPtr DAGQueryBlockInterpreter::genJoinOtherConditionAction(
...
-
-    bool swap_join_side;
-    if (kind == ASTTableJoin::Kind::Cross_Right)
-        swap_join_side = true;
-    else if (kind == ASTTableJoin::Kind::Cross_Left)
-        swap_join_side = false;
-    else
-        swap_join_side = join.inner_idx() == 0;
+    auto [kind, build_side_index] = JoinInterpreterHelper::getJoinKindAndBuildSideIndex(join);
+    assert(build_side_index == 1 || build_side_index == 0);
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:481]: use `RUNTIME_ASSSERT`?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -568,17 +497,17 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
-    for (auto const & p : input_streams_vec[0][0]->getHeader().getNamesAndTypesList())
+    for (auto const & p : input_streams_vec[0][0]->getHeader())
     {
         join_output_columns.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
         columns_for_other_join_filter.emplace_back(p.name, make_nullable ? makeNullable(p.type) : p.type);
         column_set_for_other_join_filter.emplace(p.name);
     }
     make_nullable = join.join_type() == tipb::JoinType::TypeLeftOuterJoin;
-    for (auto const & p : input_streams_vec[1][0]->getHeader().getNamesAndTypesList())
+    for (auto const & p : input_streams_vec[1][0]->getHeader())
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:510]: could you give `input_streams_vec[1][0]` a name?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 3]
```diff
@@ -568,17 +497,17 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
     /// append the extra columns afterwards. In order to figure out
     /// whether a given column is already in the column vector or
     /// not quickly, we use another set to store the column names
-    std::vector<NameAndTypePair> columns_for_other_join_filter;
+    NamesAndTypes columns_for_other_join_filter;
     std::unordered_set<String> column_set_for_other_join_filter;
     bool make_nullable = join.join_type() == tipb::JoinType::TypeRightOuterJoin;
-    for (auto const & p : input_streams_vec[0][0]->getHeader().getNamesAndTypesList())
+    for (auto const & p : input_streams_vec[0][0]->getHeader())
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:503]: could you give `input_streams_vec[0][0]` a name?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 4]
```diff
@@ -615,55 +544,41 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
-            {
-                if (join.probe_types(i).collate() != join.build_types(i).collate())
-                    throw TiFlashException("Join with different collators on the join key", Errors::Coprocessor::BadRequest);
-                collators.push_back(getCollatorFromFieldType(join.probe_types(i)));
-            }
-            else
-                collators.push_back(nullptr);
-        }
+    DataTypes join_key_types = JoinInterpreterHelper::getJoinKeyTypes(join);
+    TiDB::TiDBCollators collators = JoinInterpreterHelper::getJoinKeyCollators(join, join_key_types);
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:548]: auto?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 5]
```diff
@@ -615,55 +544,41 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
         swap_join_side ? join.right_conditions() : join.left_conditions(),
         left_filter_column_name);
 
-    prepareJoin(
+    auto build_side_prepare_actions = prepareJoin(
         swap_join_side ? join.left_join_keys() : join.right_join_keys(),
         join_key_types,
-        right_pipeline,
+        build_pipeline,
         right_key_names,
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:569]: will you replace all left/right by probe/build? mixed using two kinds of names will confuse the reader.

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 6]
```diff
@@ -697,33 +610,39 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
 
     recordJoinExecuteInfo(swap_join_side ? 0 : 1, join_ptr);
 
+    /// build side streams
+    build_pipeline.transform([&](auto & stream) { stream = std::make_shared<ExpressionBlockInputStream>(stream, build_side_prepare_actions, log->identifier()); });
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:614]: check if `build_side_prepare_actions` is null before transformation?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 7]
```diff
@@ -697,33 +610,39 @@ void DAGQueryBlockInterpreter::handleJoin(const tipb::Join & join, DAGPipeline &
...
-    right_query.source = right_pipeline.firstStream();
+    right_query.source = build_pipeline.firstStream();
     right_query.join = join_ptr;
     right_query.join->setSampleBlock(right_query.source->getHeader());
+    right_query.join->setBuildConcurrencyAndInitMap(settings.join_concurrent_build ? std::min(max_streams, build_pipeline.streams.size()) : 1);
 
-    std::vector<NameAndTypePair> source_columns;
-    for (const auto & p : left_pipeline.streams[0]->getHeader().getNamesAndTypesList())
+    /// probe side streams
+    probe_pipeline.transform([&](auto & stream) { stream = std::make_shared<ExpressionBlockInputStream>(stream, probe_side_prepare_actions, log->identifier()); });
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:627]: ditto

File: dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    /// in DAG request, inner part is the build side, however for TiFlash implementation,
+    /// the build side must be the right side, so need to swap the join side if needed
+    /// 1. for (cross) inner join, there is no problem in this swap.
+    /// 2. for (cross) semi/anti-semi join, the build side is always right, needn't swap.
+    /// 3. for non-cross left/right join, there is no problem in this swap.
+    /// 4. for cross left join, the build side is always right, needn't and can't swap.
+    /// 5. for cross right join, the build side is always left, so it will always swap and change to cross left join.
+    /// note that whatever the build side is, we can't support cross-right join now.
+
+    size_t build_side_index;
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp:62]: init?

File: dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    else
+    {
+        build_side_index = join.inner_idx();
+        if (build_side_index == 0)
+        {
+            if (kind == ASTTableJoin::Kind::Left)
+                kind = ASTTableJoin::Kind::Right;
+            else if (kind == ASTTableJoin::Kind::Right)
+                kind = ASTTableJoin::Kind::Left;
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp:80]: else?

File: dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        }
```
Comment (fuzhe1989) [inline dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp:81]: else?

File: dbms/src/Interpreters/Join.cpp [hunk 11]
```diff
@@ -127,15 +123,27 @@ Join::Join(
             strictness = ASTTableJoin::Strictness::All;
         }
     }
+    if (!left_filter_column.empty() && !isLeftJoin(kind))
+        throw Exception("Not supported: non left join with left conditions");
+    if (!right_filter_column.empty() && !isRightJoin(kind))
+        throw Exception("Not supported: non right join with right conditions");
```
Comment (fuzhe1989) [inline dbms/src/Interpreters/Join.cpp:129]: if this is due to coding mistakes, how about assert?

File: dbms/src/Interpreters/Join.cpp [hunk 12]
```diff
@@ -103,7 +102,6 @@ Join::Join(
     , key_names_left(key_names_left_)
     , key_names_right(key_names_right_)
     , use_nulls(use_nulls_)
-    , build_concurrency(std::max(1, build_concurrency_))
```
Comment (fuzhe1989) [inline dbms/src/Interpreters/Join.cpp:106]: init `build_concurrency` to 0?

File: dbms/src/Interpreters/Join.cpp [hunk 13]
```diff
@@ -127,15 +123,27 @@ Join::Join(
...
     }
+    if (!left_filter_column.empty() && !isLeftJoin(kind))
+        throw Exception("Not supported: non left join with left conditions");
+    if (!right_filter_column.empty() && !isRightJoin(kind))
+        throw Exception("Not supported: non right join with right conditions");
+}
+
+void Join::setBuildConcurrencyAndInitMap(size_t build_concurrency_)
+{
+    build_concurrency = std::max(1, build_concurrency_);
```
Comment (fuzhe1989) [inline dbms/src/Interpreters/Join.cpp:134]: assert `build_concurrency == 0 && pools.empty()`? how do you guarantee `setBuildConcurrencyAndInitMap` will and only be called once.

File: dbms/src/Interpreters/Join.cpp [hunk 14]
```diff
@@ -412,20 +420,22 @@ void Join::setSampleBlock(const Block & block)
...
-    /// Choose data structure to use for JOIN.
-    init(chooseMethod(key_columns, key_sizes));
+            /// We will join only keys, where all components are not NULL.
+            if (key_columns[i]->isColumnNullable())
+                key_columns[i] = &static_cast<const ColumnNullable &>(*key_columns[i]).getNestedColumn();
+        }
+
+        // init sample key_columns from sample block.
+        sample_key_columns = std::move(key_columns);
+    }
```
Comment (fuzhe1989) [inline dbms/src/Interpreters/Join.cpp:438]: better to be wrapped as a function

File: dbms/src/Interpreters/Join.h [hunk 15]
```diff
@@ -157,6 +156,11 @@ class Join
     bool isBuildSetExceeded() const { return build_set_exceeded.load(); }
     size_t getNotJoinedStreamConcurrency() const { return build_concurrency; };
 
+    /** Set Join build concurrency and init hash map.
+       * You must call this method before subsequent calls to insertFromBlock and setSampleBlock.
+       */
+    void setBuildConcurrencyAndInitMap(size_t build_concurrency_);
```
Comment (fuzhe1989) [inline dbms/src/Interpreters/Join.h:162]: > You must call this method before subsequent calls to insertFromBlock and setSampleBlock.

how to guarantee this?

File: dbms/src/Interpreters/Join.cpp (hunk 11 above)
Comment (SeaRise) [reply dbms/src/Interpreters/Join.cpp]: it seems to be due to an illegal input.

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp (hunk 5 above)
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp]: no.

**Pull request closed without merging**